onnxruntime/core/optimizer/graph_transformer_utils.cc

```diff
@@ -0,0 +1,109 @@
+
+#include "core/optimizer/graph_transformer_utils.h"
+#include "core/optimizer/identity_elimination.h"
+#include "core/optimizer/slice_elimination.h"
+#include "core/optimizer/conv_mul_fusion.h"
+#include "core/optimizer/conv_bn_fusion.h"
+#include "core/optimizer/conv_add_fusion.h"
+#include "core/optimizer/unsqueeze_elimination.h"
+
+namespace onnxruntime {
+
+namespace transformerutils {
+
+Status ValidateTransformerLevel(unsigned int level) {
+  auto t_level = (unsigned int)std::pow(2, level);
+  if (t_level < TransformerLevel::MaxTransformerLevel) {
+    return Status::OK();
+  }
+  return Status(common::ONNXRUNTIME, common::FAIL, "Unsupported level " + level);
+}
+
+void SetTransformerContext(const uint32_t& level, uint32_t& levels_enabled, 
+                           std::vector<TransformerLevel>* all_levels) {
+
+  ORT_ENFORCE(ValidateTransformerLevel(level).IsOK(),
+              "Unsupported transformer level specified", level);
+
+  levels_enabled = TransformerLevel::Default;
+  if (level == 1) {
+    levels_enabled |= TransformerLevel::Level1;
+  } else if (level == 2) {
+    levels_enabled |= TransformerLevel::Level1 | TransformerLevel::Level2;
+  }
+
+  if (all_levels != nullptr) {
+    all_levels->push_back(TransformerLevel::Level2);
+    all_levels->push_back(TransformerLevel::Level1);
+  }
+}
+
+std::vector<std::unique_ptr<RewriteRule>> GenerateRewriteRules(const TransformerLevel& level, 
+                                                               const std::vector<std::string>* custom_list) {
+  std::vector<std::unique_ptr<RewriteRule>> rules;
+  switch (level) {
+    case TransformerLevel::Level1:
+      rules.push_back(std::make_unique<EliminateIdentity>());
+      rules.push_back(std::make_unique<EliminateSlice>());
+      break;
+
+    case TransformerLevel::Level2:
+      break;
+    default:
+      ORT_ENFORCE(false, "Unsupported level" + level);
+  }
+
+  if (custom_list != nullptr && !custom_list->empty()) {
+    std::vector<std::unique_ptr<RewriteRule>> filtered_list;
+    for (const auto& rule_name : *custom_list) {
+      std::for_each(rules.begin(), rules.end(), [&](std::unique_ptr<RewriteRule>& item) {
+        if((item != nullptr) && (item->Name() == rule_name)) {
+          filtered_list.push_back(std::move(item));
```

Comment:
You should be the one defining the order of rules or is the custom_list doing that?
If it is the former: make the custom_list a set and then just go over the rules you registered just before and remove those that dont appear in your set.

Reply:
order does not matter because we iteratively apply the rules....

regarding why I did it this way... since custom list size in most cases is smaller or equal to generated rules so I decided to iterate over custom and select rules as opposed to iterate over rules and remove them... in this case since I am not copying the objects it should not matter.... what I have is not going to be inefficient 
+ we don't have as many transformers right now to make a perf difference in this loop
+ we only do this once during initialization


---
In reply to: [265757590](https://github.com/Microsoft/onnxruntime/pull/608#discussion_r265757590) [](ancestors = 265757590)

Reply:
But for each rule in the filtered_list, you will be iterating over all rules. So this will be strictly more iterations, right?
I do agree that over-optimizing this piece of code is probably not worth it, but it will also lead to simpler code.

Reply:
We can keep it this way if we want to impose the order of rules_to_enable. Otherwise, a set would be more efficient.
Assuming we want the former, we can leave it as is.